### PR TITLE
[WIP] Single malloc for all parser state

### PR DESCRIPTION
**Pull request by jkeiser**
This patch mallocs all parser internal state memory in one big heap, and doles out pointers to it on demand. This does not consolidate *document* memory in any way; that will be separate, partly because we want people to be able to steal the document pointer in a single copy, which has a reasonable chance of improving the Javascript binding speed.

This has two purposes:

1. See if we can reduce malloc overhead of creating a parser.
2. Move towards a world where the architecture-specific implementation mallocs its own internal state, so it can be customized and owned there. I expect this to make custom implementations easier to work on, as well as make streaming easier (because then we can allocate enough room to store stage 1/2 intermediate state, which is implementation-specific).

## Layout

Because the formulas for ret_address and containing_scope are simpler, I laid it out thus:

```
ret_address: void*[max_depth];
containing_scope: scope_descriptor[max_depth];
structural_indexes: uint32_t[ROUNDUP_N(capacity, 64) + 2 + 7)];
```

## Performance

I have attempted zero optimization here (for example, I expect that storing the pointers in a local variable will help recover some or all lost speed ...), and am marking this WIP.

I see on the order of 3% degradation with the patch as written. Putting this up for folks to have a look, and tell me if you see any red flags or optimization opportunities.

### This PR: jkeiser/parser-malloc

```
All Stages
|    Speed        :  23.8853 ns per block ( 84.44%) -   0.3732 ns per byte -   4.2651 ns per structural -   2.6793 GB/s
|- Stage 1
|    Speed        :  11.3498 ns per block ( 40.13%) -   0.1774 ns per byte -   2.0267 ns per structural -   5.6385 GB/s
|- Stage 2
|    Speed        :  12.4747 ns per block ( 44.10%) -   0.1949 ns per byte -   2.2275 ns per structural -   5.1301 GB/s

4242.7 documents parsed per second (best)
```

### master

```
All Stages
|    Speed        :  23.1455 ns per block ( 87.63%) -   0.3617 ns per byte -   4.1330 ns per structural -   2.7650 GB/s
|- Stage 1
|    Speed        :  11.0965 ns per block ( 42.01%) -   0.1734 ns per byte -   1.9814 ns per structural -   5.7673 GB/s
|- Stage 2
|    Speed        :  12.0085 ns per block ( 45.46%) -   0.1876 ns per byte -   2.1443 ns per structural -   5.3292 GB/s

4378.3 documents parsed per second (best)
```

### CODE REVIEW

**Comment by jkeiser** [comment]
@lemire [this patch](https://github.com/simdjson/simdjson/pull/862/commits/bca9bc0de272fec0ad465a6034dbdc1fc6bd97c5) recovers stage 1 performance entirely for me. I can't for the life of me imagine why. The variable in question is only used before and after the hot loop, not during it ...

**Comment by lemire** [comment]
On my todo.

**Comment by jkeiser** [comment]
There is still a possible 2% degradation in stage 1 after making the above change (it's hard to tell when they are that small).

Stage 2's 5% degradation is more mysterious; I tried copying ret_address in the same way and got a 5% *loss*. I didn't try to localize containing_scope(), but it requires no calculation so I expect no gain. structural_indexes() is only used once during stage 2 (it's already copied).

**Comment by jkeiser** [comment]
@lemire I just pushed a LOT of extra changes, the intent of which was to move all document writing state into the stage 2 code. I'm super excited about that.

After that, this change overall improves performance by between 5-10% on my machine. I *think* the only thing that *actually* improved performance was the checkin where I moved on_number_* into stage 2, and it's possible we could get that improvement without any other changes (I'll check). I had to retool parse_large_integer to keep the call to on_number_* inline, leaving the actual integer parsing (or most of it) in a never_inline function that still does most of the work (if you just make that function really_inline you pay a price, even though the branch is unlikely). This was to avoid having the stage 2 parser passed as an object to a non-inline function, which in the past has prevented the optimizer from enregistering it (destructuring it? What is the right term for the optimization where the compiler turns a struct into a series of registers instead of a contiguous memory layout?).

**Comment by jkeiser** [comment]
Um. Never mind, that last bit is wrong. It turns out that it's important to pay attention to the most significant digit, and that 3500 is *lower* than 4300.

**Comment by jkeiser** [comment]
Yeah, the number parsing change didn't improve by 300 docs/s, it dropped 700 docs/s :) OK, give me a sec to fight with that code.

**Comment by jkeiser** [comment]
@lemire given all the above, I've reverted it back to just the malloc bit. At this point, we're at the "it makes things slightly slower, and the second commit recovers a lot of the stage 1 performance, but I don't understand why yet."

**Comment by lemire** [comment]
@jkeiser There is a small but clear perf. degradation, I think. It may be acceptable.

**Comment by lemire** [comment]
I am seeing a difference only in stage 2.

**Comment by lemire** [comment]
@jkeiser I don't understand the motivation for playing with ret_address. That thing could even be allocated on stack during stage 2. My guess is that all of the perf. degradation is tied with the changes you made to ret_address.

I am not so worried about the perf difference, as I am curious as to why you'd change ret_address. This is very narrowly related to one piece of code.

**Comment by jkeiser** [comment]
> @jkeiser I don't understand the motivation for playing with ret_address. That thing could even be allocated on stack during stage 2. My guess is that all of the perf. degradation is tied with the changes you made to ret_address.
> 
> I am not so worried about the perf difference, as I am curious as to why you'd change ret_address. This is very narrowly related to one piece of code.

I actually don't think I made any significant change to it besides mallocing it in the same spot as all other internal parser state--is that what you're talking about?

Oh, is this because it now has a [formula calculating it](https://github.com/simdjson/simdjson/pull/862/files#diff-8f4ffd1829b7f8d358af98574a03656cR547)? I believe I tried making a pointer to it at the beginning of stage 2 and it not making a difference, but I've made at least one change since then, so let me try again.

**Comment by jkeiser** [comment]
@lemire I just tried keeping ret_address a pointer in dom::parser, and then initializing that pointer immediately after malloc. It made zero difference. I believe the optimizer figures this thing out.

The difference here is damn near the level of noise that stage 2 normally operates at (2%), but I'm still bothered by it. I'm going to try another thing or two.

**Comment by jkeiser** [comment]
@lemire I just fixed it so that all parser code is identical, and allocate() assigns the three pointers right after malloc. I still see the small degradation, and it puzzles me more now. I'm wondering if it has to do with the way the memory is getting faulted in / initialized ... now instead of starting at the beginning of three separate malloced chunks, we're starting at three independent parts of a single malloced chunk.

### Master

```
All Stages
|    Speed        :  23.1557 ns per block ( 84.69%) -   0.3618 ns per byte -   4.1348 ns per structural -   2.7637 GB/s
|- Stage 1
|    Speed        :  11.1066 ns per block ( 40.62%) -   0.1736 ns per byte -   1.9832 ns per structural -   5.7620 GB/s
|- Stage 2
|    Speed        :  11.9882 ns per block ( 43.84%) -   0.1873 ns per byte -   2.1407 ns per structural -   5.3383 GB/s

4376.4 documents parsed per second (best)
```

### PR

```
All Stages
|    Speed        :  23.9765 ns per block ( 81.62%) -   0.3747 ns per byte -   4.2813 ns per structural -   2.6691 GB/s
|- Stage 1
|    Speed        :  11.3296 ns per block ( 38.57%) -   0.1770 ns per byte -   2.0231 ns per structural -   5.6486 GB/s
|- Stage 2
|    Speed        :  12.5963 ns per block ( 42.88%) -   0.1968 ns per byte -   2.2492 ns per structural -   5.0806 GB/s

4226.5 documents parsed per second (best)
```

We lose 2% on stage 1, 5% on stage 2, and 3.5% overall. The malloc itself makes no difference.

This could still be justified if Windows is way better, and it might be. I'll check. But I'm super puzzled, because there is absolutely no reason the actual parser code should be *any different*. I'll check instruction counts to confirm or deny that when skylake/skylake-x are up (can't check it on Windows with our tools at the moment, even in WSL2).

**Comment by jkeiser** [comment]
Current thought: see if we can pass our enregistration-sensitive structs around less. So either make a new small thing we can pass to numberparsing (like tape_writer or something), or have numberparsing return a type and value, and have the *caller* write it out.

**Comment by lemire** [comment]
Looking at my numbers, this PR increases slightly the number of instructions and also the number of cache misses in stage 2. This suggests that this PR changes how the memory is accessed in a way that is more difficult to predict for the processor.

**Comment by jkeiser** [comment]
The increase in instruction count is the most puzzling bit to be. I'd expect a little bit of cache miss because (I'd warrant) malloc probably faults in the first few pages ... but I wouldn't expect *much*.

**Comment by jkeiser** [comment]
Had a brief theory about alignment--we're mallocing chars, which don't have to be aligned on 64-bit boundaries, but then we're accessing it as 64-bit pointers. Changing that doesn't fix any perf.

I also tried using new instead of free so we don't have to parameterize the template, but that didn't make anything faster, either.

I think I'm going to work on this from the side, and see if the stack-based depth technique I used in the fallback parser can be used here (allocate depth=16 for the function and recurse if needed). That would entirely eliminate the need to malloc scope_descriptor and ret_address, and might make things faster all by its own lonesome. Might not work, too, as stage 2 is just damned fiddly.

**Comment by jkeiser** [comment]
Hey, I have another idea though. If this is a cache access thing, we're not really using most of containing_scope and ret_address--probably no more than one page each. Maybe putting structural_indexes first would actually fix this ...

**Comment by jkeiser** [comment]
... that wasn't it, either :/

**Comment by lemire** [comment]
> I think I'm going to work on this from the side, and see if the stack-based depth technique I used in the fallback parser can be used here (allocate depth=16 for the function and recurse if needed). That would entirely eliminate the need to malloc scope_descriptor and ret_address, and might make things faster all by its own lonesome. Might not work, too, as stage 2 is just damned fiddly.

Something of the sort seems most promising to me.

**Comment by jkeiser** [comment]
Closing this to try again with a stack-based approach to ret_address and containing_scope.
